File: sdk/typescript/src/delta.test.ts
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { describe, expect, it } from "vitest";
import { applyDelta, canApplyDelta } from "./delta.js";
import type { Delta, Snapshot } from "./types.js";

describe("canApplyDelta", () => {
  it("is true when from matches the current revision", () => {
    const current: Snapshot = { version: 1, revision: 5, schemaHash: "h", values: {} };
    const delta: Delta = { type: "delta", version: 1, revision: 6, from: 5, schemaHash: "h" };

    expect(canApplyDelta(current, delta)).toBe(true);
  });

  it("is false on a from-mismatch (defensive path — caller should full-resync instead)", () => {
    const current: Snapshot = { version: 1, revision: 5, schemaHash: "h", values: {} };
    const delta: Delta = { type: "delta", version: 1, revision: 9, from: 7, schemaHash: "h" };

    expect(canApplyDelta(current, delta)).toBe(false);
  });
});

describe("applyDelta", () => {
  it("merges values.set/remove onto a copy of current.values", () => {
    const current: Snapshot = {
      version: 1,
      revision: 1,
      schemaHash: "h1",
      values: { retries: 3, timeout: 30 },
    };
    const delta: Delta = {
      type: "delta",
      version: 1,
      revision: 2,
      from: 1,
      schemaHash: "h1",
      values: { set: { retries: 5 }, remove: ["timeout"] },
    };

    const result = applyDelta(current, delta);

    expect(result).toEqual({
      version: 1,
      revision: 2,
      schemaHash: "h1",
      values: { retries: 5 },
      targeting: {},
    });
  });

  it("merges targeting.set/remove onto a copy of current.targeting", () => {
    const current: Snapshot = {
      version: 1,
      revision: 1,
      schemaHash: "h1",
      values: {},
      targeting: { beta: [{ value: false }], gamma: [{ value: true }] },
    };
    const delta: Delta = {
      type: "delta",
      version: 1,
      revision: 2,
      from: 1,
      schemaHash: "h1",
      targeting: { set: { beta: [{ value: true }] }, remove: ["gamma"] },
    };

    const result = applyDelta(current, delta);

    expect(result.targeting).toEqual({ beta: [{ value: true }] });
  });

  it("treats an absent values/targeting on the delta as no change on that side", () => {
    const current: Snapshot = {
      version: 1,
      revision: 1,
      schemaHash: "h1",
      values: { a: 1 },
      targeting: { beta: [{ value: false }] },
    };
    const delta: Delta = { type: "delta", version: 2, revision: 2, from: 1, schemaHash: "h2" };

    const result = applyDelta(current, delta);

    expect(result).toEqual({
      version: 2,
      revision: 2,
      schemaHash: "h2",
      values: { a: 1 },
      targeting: { beta: [{ value: false }] },
    });
  });

  it("treats an absent set/remove within values/targeting as no change on that side", () => {
    const current: Snapshot = {
      version: 1,
      revision: 1,
      schemaHash: "h1",
      values: { a: 1 },
      targeting: { beta: [{ value: false }] },
    };
    // values has no `remove`, targeting has no `set` — both should be no-ops on their side.
    const delta: Delta = {
      type: "delta",
      version: 1,
      revision: 2,
      from: 1,
      schemaHash: "h1",
      values: { set: { b: 2 } },
      targeting: { remove: ["beta"] },
    };

    const result = applyDelta(current, delta);

    expect(result.values).toEqual({ a: 1, b: 2 });
    expect(result.targeting).toEqual({});
  });

  it("does not mutate the input snapshot or its nested objects", () => {
    const current: Snapshot = {
      version: 1,
      revision: 1,
      schemaHash: "h1",
      values: { a: 1 },
      targeting: { beta: [{ value: false }] },
    };
    const snapshotBefore = JSON.parse(JSON.stringify(current));
    const delta: Delta = {
      type: "delta",
      version: 1,
      revision: 2,
      from: 1,
      schemaHash: "h1",
      values: { set: { a: 2, c: 3 }, remove: ["a"] },
      targeting: { set: { gamma: [{ value: true }] }, remove: ["beta"] },
    };

    applyDelta(current, delta);

    expect(current).toEqual(snapshotBefore);
  });
});

// The golden-vector contract: this file is a byte-for-byte copy of
// internal/delivery/testdata/delta-vectors.json (Task 1's brief). The TS SDK's applyDelta
// must reproduce every step recorded there — that's how the TS and Go SDKs stay pinned
// together on delta apply semantics.
//
// The vectors file carries a `_comment` key documenting its own shape (see the file itself)
// — plain JSON.parse ignores it, so no strict/`additionalProperties: false` validation is
// used here.
const vectorsPath = fileURLToPath(new URL("./__testdata__/delta-vectors.json", import.meta.url));

interface DeltaVectorsFile {
  base: Snapshot;
  steps: Array<{ delta: Delta; expected: Snapshot }>;
}

const vectors: DeltaVectorsFile = JSON.parse(readFileSync(vectorsPath, "utf8"));

describe("applyDelta (golden vectors)", () => {
  it("has at least one step", () => {
    expect(vectors.steps.length).toBeGreaterThan(0);
  });

  it("replays every step from base, matching the expected snapshot at each step", () => {
    let current = vectors.base;
    for (const step of vectors.steps) {
      expect(canApplyDelta(current, step.delta)).toBe(true);
      current = applyDelta(current, step.delta);
      expect(current).toEqual(step.expected);
    }
  });
});

File: sdk/go/delta_test.go
package knobs

import (
	"encoding/json"
	"os"
	"reflect"
	"testing"
)

func TestApplyDeltaMergesValuesAndTargeting(t *testing.T) {
	current := Snapshot{
		Version:    1,
		Revision:   5,
		SchemaHash: "h1",
		Values:     map[string]any{"a": 1.0, "b": 2.0},
		Targeting:  map[string][]Rule{"flag": {{Value: false}}},
	}
	d := Delta{
		Type:       "delta",
		Version:    2,
		Revision:   6,
		From:       5,
		SchemaHash: "h2",
		Values: FieldDelta{
			Set:    map[string]any{"a": 10.0, "c": 3.0},
			Remove: []string{"b"},
		},
		Targeting: TargetingDelta{
			Set:    map[string][]Rule{"flag": {{Value: true}}, "new": {{Value: "x"}}},
			Remove: []string{"old"},
		},
	}

	result := applyDelta(current, d)

	wantValues := map[string]any{"a": 10.0, "c": 3.0}
	if !reflect.DeepEqual(result.Values, wantValues) {
		t.Errorf("Values = %#v, want %#v", result.Values, wantValues)
	}
	wantTargeting := map[string][]Rule{"flag": {{Value: true}}, "new": {{Value: "x"}}}
	if !reflect.DeepEqual(result.Targeting, wantTargeting) {
		t.Errorf("Targeting = %#v, want %#v", result.Targeting, wantTargeting)
	}
	if result.Version != 2 || result.Revision != 6 || result.SchemaHash != "h2" {
		t.Errorf("Version/Revision/SchemaHash = %d/%d/%s, want 2/6/h2", result.Version, result.Revision, result.SchemaHash)
	}

	// current must be untouched by the call...
	if len(current.Values) != 2 || current.Values["a"] != 1.0 || current.Values["b"] != 2.0 {
		t.Errorf("applyDelta mutated current.Values: %#v", current.Values)
	}
	if len(current.Targeting) != 1 || len(current.Targeting["flag"]) != 1 || current.Targeting["flag"][0].Value != false {
		t.Errorf("applyDelta mutated current.Targeting: %#v", current.Targeting)
	}

	// ...and result holds its own maps, not aliases of current's.
	result.Values["z"] = "mutated-after-the-fact"
	if _, ok := current.Values["z"]; ok {
		t.Error("result.Values shares backing storage with current.Values")
	}
	result.Targeting["z"] = []Rule{{Value: "mutated-after-the-fact"}}
	if _, ok := current.Targeting["z"]; ok {
		t.Error("result.Targeting shares backing storage with current.Targeting")
	}
}

func TestApplyDeltaAbsentSetRemoveIsNoChange(t *testing.T) {
	current := Snapshot{
		Version:    1,
		Revision:   5,
		SchemaHash: "h1",
		Values:     map[string]any{"a": 1.0},
		Targeting:  map[string][]Rule{"flag": {{Value: false}}},
	}
	// Values and Targeting are their zero value (FieldDelta{}/TargetingDelta{}) —
	// absent Set/Remove on the wire, which must mean "no change" on that side.
	d := Delta{Type: "delta", Version: 1, Revision: 6, From: 5, SchemaHash: "h1"}

	result := applyDelta(current, d)

	if !reflect.DeepEqual(result.Values, current.Values) {
		t.Errorf("Values = %#v, want unchanged %#v", result.Values, current.Values)
	}
	if !reflect.DeepEqual(result.Targeting, current.Targeting) {
		t.Errorf("Targeting = %#v, want unchanged %#v", result.Targeting, current.Targeting)
	}
	if result.Revision != 6 || result.SchemaHash != "h1" {
		t.Errorf("Revision/SchemaHash = %d/%s, want 6/h1", result.Revision, result.SchemaHash)
	}
}

func TestApplyDeltaDoesNotItselfCheckFrom(t *testing.T) {
	// applyDelta is documented as pure and not responsible for the
	// from-vs-current.Revision gate — that's the caller's job
	// (Client.applyDeltaFrame decides whether to call this at all, and
	// falls back to a full resync when From doesn't chain). A mismatched
	// From here must still apply mechanically, so this pins that contract:
	// the mismatch handling lives at the call site, not in applyDelta.
	current := Snapshot{
		Revision: 5,
		Values:   map[string]any{"a": 1.0},
	}
	d := Delta{Revision: 99, From: 12345, Values: FieldDelta{Set: map[string]any{"a": 2.0}}}

	result := applyDelta(current, d)

	if result.Revision != 99 || result.Values["a"] != 2.0 {
		t.Errorf("applyDelta result = %+v, want it to apply mechanically regardless of From", result)
	}
}

// deltaVectors mirrors the shape documented by delta-vectors.json's
// "_comment": {base: Snapshot, steps: [{delta: Delta, expected: Snapshot}, ...]}.
// json.Unmarshal ignores the unknown "_comment" key by default — this must
// NOT use DisallowUnknownFields.
type deltaVectors struct {
	Base  Snapshot `json:"base"`
	Steps []struct {
		Delta    Delta    `json:"delta"`
		Expected Snapshot `json:"expected"`
	} `json:"steps"`
}

func TestApplyDeltaMatchesGoldenVectors(t *testing.T) {
	data, err := os.ReadFile("testdata/delta-vectors.json")
	if err != nil {
		t.Fatalf("reading testdata/delta-vectors.json: %v", err)
	}

	var vectors deltaVectors
	if err := json.Unmarshal(data, &vectors); err != nil {
		t.Fatalf("unmarshalling delta-vectors.json: %v", err)
	}
	if len(vectors.Steps) == 0 {
		t.Fatal("delta-vectors.json has no steps")
	}

	current := vectors.Base
	for i, step := range vectors.Steps {
		current = applyDelta(current, step.Delta)

		if !reflect.DeepEqual(current.Values, step.Expected.Values) {
			t.Errorf("step %d: Values = %#v, want %#v", i, current.Values, step.Expected.Values)
		}
		if !reflect.DeepEqual(current.Targeting, step.Expected.Targeting) {
			t.Errorf("step %d: Targeting = %#v, want %#v", i, current.Targeting, step.Expected.Targeting)
		}
		if current.Version != step.Expected.Version {
			t.Errorf("step %d: Version = %d, want %d", i, current.Version, step.Expected.Version)
		}
		if current.Revision != step.Expected.Revision {
			t.Errorf("step %d: Revision = %d, want %d", i, current.Revision, step.Expected.Revision)
		}
		if current.SchemaHash != step.Expected.SchemaHash {
			t.Errorf("step %d: SchemaHash = %q, want %q", i, current.SchemaHash, step.Expected.SchemaHash)
		}
	}
}
